fusequery/query/src/api/http_service.rs

```diff
@@ -2,26 +2,107 @@
 //
 // SPDX-License-Identifier: Apache-2.0.
 
-use anyhow::Result;
+use std::net::SocketAddr;
+use std::net::ToSocketAddrs;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::time::Duration;
+use std::time::Instant;
+
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_exception::ToErrorCode;
+use common_infallible::Mutex;
+use tokio::sync::mpsc::Sender;
+use tokio::sync::Notify;
 
 use crate::api::http::router::Router;
 use crate::clusters::ClusterRef;
 use crate::configs::Config;
+use crate::servers::AbortableServer;
+use crate::servers::AbortableService;
+use crate::servers::Elapsed;
 
 pub struct HttpService {
     cfg: Config,
     cluster: ClusterRef,
+    aborted: Arc<AtomicBool>,
+    abort_handle: Mutex<Option<Sender<()>>>,
+    aborted_notify: Arc<Notify>,
 }
 
 impl HttpService {
-    pub fn create(cfg: Config, cluster: ClusterRef) -> Self {
-        HttpService { cfg, cluster }
+    pub fn create(cfg: Config, cluster: ClusterRef) -> AbortableServer {
+        Arc::new(HttpService {
+            cfg,
+            cluster,
+            aborted: Arc::new(AtomicBool::new(false)),
+            abort_handle: Mutex::new(None),
+            aborted_notify: Arc::new(Notify::new()),
+        })
     }
+}
+
+#[async_trait::async_trait]
+impl AbortableService<(String, u16), SocketAddr> for HttpService {
+    fn abort(&self, _force: bool) -> Result<()> {
+        if let Some(abort_handle) = self.abort_handle.lock().take() {
+            futures::executor::block_on(abort_handle.send(()))
+                .map_err_to_code(ErrorCode::LogicalError, || "Cannot abort HttpService.")?;
+        }
 
-    pub async fn make_server(&self) -> Result<()> {
-        let address = self.cfg.http_api_address.parse::<std::net::SocketAddr>()?;
-        let router = Router::create(self.cfg.clone(), self.cluster.clone());
-        warp::serve(router.router()?).run(address).await;
         Ok(())
     }
+
+    async fn start(&self, args: (String, u16)) -> Result<SocketAddr> {
+        let router = Router::create(self.cfg.clone(), self.cluster.clone());
+        let server = warp::serve(router.router()?);
+
+        let (tx, mut rx) = tokio::sync::mpsc::channel(1);
```

Comment:
oneshot is better ? 
https://docs.rs/tokio/0.1.22/tokio/sync/oneshot/index.html